from PyQt5 import uic
from datetime import datetime
from PyQt5.QtWidgets import QDialog
from PyQt5.QtCore import pyqtSignal, Qt


class TaskWidget(QDialog):
    taskClicked = pyqtSignal(int)

    def __init__(self, task):
        super().__init__()

        self.task_id = task['id']
        self.title = task['title']
        self.assignee = task['assignee']['full_name']
        self.priority = task['priority']
        self.deadline = task['deadline']

        uic.loadUi('client/http_client/ui/main/TaskWidget.ui', self)

        with open('client/http_client/ui/styles/task.qss', 'r', encoding='utf-8') as f:
            self.setStyleSheet(f.read())

        self.setup_task_info()
        self.setMouseTracking(True)

    def setup_task_info(self):
        self.label_title.setText(f"üìå–ù–∞–∑–≤–∞–Ω–∏–µ: {self.title}")
        self.label_assignee.setText(f"üë§–†–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫: {self.assignee}")

        priority_text, color = self.get_priority_text(self.priority)
        self.label_priority.setText(f"üö®–ü—Ä–∏–æ—Ä–∏—Ç–µ—Ç: {priority_text}")
        self.label_priority.setStyleSheet(f"color: {color};")

        self.label_deadline.setText(f"üìÖ–î–µ–¥–ª–∞–π–Ω: {self.format_datetime(self.deadline)}")

    def get_priority_text(self, priority_value):
        try:
            priority = float(priority_value)
            if priority <= 0.20:
                return "–æ—á–µ–Ω—å –Ω–∏–∑–∫–∏–π üü¢", "#90ee90"
            elif 0.21 <= priority <= 0.35:
                return "–Ω–∏–∑–∫–∏–π üü¢", "#008000"
            elif 0.36 <= priority <= 0.55:
                return "—Å—Ä–µ–¥–Ω–∏–π üü°", "#cccc00"
            elif 0.56 <= priority <= 0.75:
                return "–≤—ã—Å–æ–∫–∏–π üî¥", "#ff6666"
            elif 0.76 < priority <= 0.99:
                return "–æ—á–µ–Ω—å –≤—ã—Å–æ–∫–∏–π üî¥", "#8b0000"
            return "–Ω–µ –æ–ø—Ä–µ–¥–µ–ª—ë–Ω", "#000000"
        except:
            return "–Ω–µ –æ–ø—Ä–µ–¥–µ–ª—ë–Ω", "#000000"

    def format_datetime(self, datetime_str):
        try:
            if not datetime_str:
                return ""

            datetime_str = datetime_str.rstrip("Z")

            for fmt in ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%dT%H:%M:%S.%f"):
                try:
                    dt = datetime.strptime(datetime_str, fmt)
                    return dt.strftime("%d.%m.%Y %H:%M")
                except ValueError:
                    continue

            return datetime_str
        except Exception as e:
            return datetime_str

    def mousePressEvent(self, event):
        if event.button() == Qt.LeftButton:
            self.taskClicked.emit(self.task_id)
